sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -152,26 +153,54 @@ class GpuSingleDirectoryDataWriter(
   override def write(batch: ColumnarBatch): Unit = {
     var needToCloseBatch = true
     try {
-      // TODO: This does not handle batch splitting to make sure maxRecordsPerFile is not exceeded.
-      if (description.maxRecordsPerFile > 0 && recordsInFile >= description.maxRecordsPerFile) {
-        fileCounter += 1
-        assert(fileCounter < MAX_FILE_COUNTER,
-          s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
+      withResource(GpuColumnVector.from(batch)) {table => 
+        val numRows = table.getRowCount()
+        val maxRecordsPerFile = description.maxRecordsPerFile.toInt
+        val fillOldFile = recordsInFile < maxRecordsPerFile
+        val splitIndexes = (maxRecordsPerFile > 0, fillOldFile) match {
+          case (false, _) => IndexedSeq.empty
+          case (true, false) =>
+            (1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => i * maxRecordsPerFile)
+          case (true, true) => {
+            val filledUp = maxRecordsPerFile - recordsInFile.toInt
+            val remain = numRows - filledUp
+            (0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => filledUp + i * maxRecordsPerFile)
+          }
+        }
 
-        newOutputWriter()
+        val dataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType()).toArray
+
+        var needNewWriter = !fillOldFile
```

Comment:
This is the only real use of `fillOldFile` which means we should just ditch that name and create needNewWriter directly as `recordsInFile >= maxRecordsPerFile`.

Reply:
The `fillOldFile` is also used here:
```scala
        val splitIndexes = (maxRecordsPerFile > 0, fillOldFile) match {
          ...
```